build_runner/lib/src/build_script_generate/builder_ordering.dart

```diff
@@ -0,0 +1,40 @@
+import 'package:build_config/build_config.dart';
+
+import '../package_graph/dependency_ordering.dart';
+
+/// Put [builders] into an order such that any builder which specifies
+/// [BuilderDefinition.requiredInputs] will come after any builder which
+/// produces a desired output.
+///
+/// All builders which don't specify required inputs will come first, followed
+/// by the builders with required inputs. If there is a cycle in required inputs
+/// this will throw.
+Iterable<BuilderDefinition> findBuilderOrder(
+    Iterable<BuilderDefinition> builders) {
+  final result = <BuilderDefinition>[];
+  result.addAll(builders.where((b) => b.requiredInputs.isEmpty));
+  final toOrder = builders.where((b) => b.requiredInputs.isNotEmpty);
+  result.addAll(_findOrder(toOrder));
+  return result;
+}
+
+List<BuilderDefinition> _findOrder(Iterable<BuilderDefinition> builders) {
+  Iterable<BuilderDefinition> dependencies(BuilderDefinition parent) =>
+      builders.where((child) => _hasDependency(parent, child));
+  var components = stronglyConnectedComponents<String, BuilderDefinition>(
+      builders, _builderKey, dependencies);
+  return components.map((component) {
+    if (component.length > 1) {
+      throw new ArgumentError('Required input cycle for ${component.toList()}');
+    }
+    return component.single;
+  }).toList();
+}
+
+bool _hasDependency(BuilderDefinition parent, BuilderDefinition child) {
```

Comment:
nit: consider `_hasInputDependency` or something so its a bit more clear what a dependency means in this context

Reply:
Done